primer/src/Primer/Eval/NormalOrder.hs

```diff
@@ -92,29 +85,34 @@ data RedexWithContext
 data ViewLet = ViewLet
   { bindingVL :: LetBinding
   -- ^ the binding itself
-  , bodyVL :: Accum Cxt (Dir, ExprZ)
-  -- ^ the body (i.e. after the `in`)
-  , typeChildrenVL :: [Accum Cxt TypeZ]
-  -- ^ any non-body type children
-  , termChildrenVL :: [Accum Cxt (Dir, ExprZ)]
-  -- ^ any non-body term children (i.e. rhs of the binding)
+  , bodyVL :: (Maybe LetBinding, (Dir, ExprZ))
+  -- ^ the body (i.e. after the `in`), along with the newly in-scope bindings (this will always be @Just bindingVL@, but record it here for consistency)
+  -- , bodyVL::Accum Cxt (Dir, ExprZ) -- ^ the body (i.e. after the `in`)
+  , typeChildrenVL :: [(Maybe LetBinding, TypeZ)]
+  -- ^ any non-body type children, along with (potentially) the newly in-scope binding
+  , termChildrenVL :: [(Maybe LetBinding, (Dir, ExprZ))]
+  -- ^ any non-body term children (i.e. rhs of the binding), along with (potentially) the newly in-scope binding
+  -- , typeChildrenVL::[Accum Cxt TypeZ] -- ^ any non-body type children
+  -- , termChildrenVL ::[Accum Cxt (Dir, ExprZ)] -- ^ any non-body term children (i.e. rhs of the binding)
   }
 viewLet :: (Dir, ExprZ) -> Maybe ViewLet
 viewLet dez@(_, ez) = case (target ez, exprChildren dez) of
-  (Let _ x e _b, [ez', bz]) -> Just (ViewLet (LetBind x e) bz [] [ez'])
-  (Letrec _ x e ty _b, [ez', bz]) -> Just (ViewLet (LetrecBind x e ty) bz tz [ez'])
-  (LetType _ a ty _b, [bz]) -> bz `seq` Just (ViewLet (LetTyBind $ LetTypeBind a ty) bz tz [])
+  (Let _ x e _b, [ez', bz]) -> let l = LetBind x e in Just (ViewLet l (Just l, bz) [] [(Nothing, ez')])
+  (Letrec _ x e ty _b, [ez', bz]) -> let l = LetrecBind x e ty in Just (ViewLet l (Just l, bz) tz [(Just l, ez')])
+  (LetType _ a ty _b, [bz]) -> bz `seq` let l = LetTyBind $ LetTypeBind a ty in Just (ViewLet l (Just l, bz) tz [])
   _ -> Nothing
   where
-    tz :: [Accum Cxt TypeZ]
+    tz :: [(Maybe LetBinding, TypeZ)]
     -- as with focusType', we don't need to bind anything here
-    tz = maybe [] ((: []) . pure) $ focusType ez
+    tz = maybe [] ((: []) . (Nothing,)) $ focusType ez
 
 -- | This is similar to 'foldMap', with a few differences:
 -- - 'Expr' is not foldable
 -- - We target every subexpression and also every (nested) subtype (e.g. in an annotation)
 -- - We keep track of context and directionality (necessitating an extra 'Dir' argument, for "what directional context this whole expression is in")
+-- TODO: the "context" is just of the immediately-enclosing lets, which are the only ones that may "cross a binder"
 -- - We handle @let@s specially, since we need to handle them differently when finding the normal order redex.
+-- TODO: do we still "handle lets specially", i.e. do we ever set subst = Just ...?
```

Comment:
This is some of the simplifications I alluded to earlier. I plan to do these simple ones shortly